### PR TITLE
convert Null values of is_sticky to 0 to avoid sort problems

### PR DESCRIPTION
**Pull request by mostafa-hisham**
should help fixes #[33](https://github.com/Strategy11/business-directory-ratings/issues/33) when there is sticky listing

### CODE REVIEW

**Review by garretlaxton — APPROVED** [review]
This looks good as well!

**Review by stephywells — APPROVED** [review]
Looks good!

Just curious @mostafa-hisham . Have you checked the query speed to see if this affects it?

**Comment by stephywells** [comment]
@mostafa-hisham Thanks for checking! Is there a better way of doing this that won't add extra load on the DB? What happens if we run the SELECT sticky query separately?

**Comment by stephywells** [comment]
@mostafa-hisham I was thinking something like:
`$sticky_posts = "SELECT listing_id FROM {$wpdb->prefix}wpbdp_listings WHERE is_sticky=1";`

And then in the main query, we would add this to the WHERE query:
```
if ( $sticky_posts ) {
   ' AND in (' . implode( $sticky_posts, ',' ) . ')';
}
```

I think this might cut down the time, but I'm not positive.

**Comment by stephywells** [comment]
Oh, right. But I think the same logic would apply, but using the ids for sorting. Right?

File: includes/class-query-integration.php [hunk 1]
```diff
@@ -249,13 +249,11 @@ public function posts_clauses( $pieces, $query ) {
 		$pieces = apply_filters( 'wpbdp_query_clauses', $pieces, $query );
 
 		// Sticky listings.
-		$is_sticky_query = "(SELECT is_sticky FROM {$wpdb->prefix}wpbdp_listings wls WHERE wls.listing_id = {$wpdb->posts}.ID LIMIT 1) AS wpbdp_is_sticky";
-
+		$sticky_ids_str = $this->get_sticky_listing_ids();
 		if ( in_array( wpbdp_current_view(), wpbdp_get_option( 'prevent-sticky-on-directory-view' ), true ) ) {
-			$is_sticky_query = '';
+			$sticky_ids_str = '';
```
Comment (stephywells) [inline includes/class-query-integration.php:254]: Can we reverse the logic here, so we only run the DB query when we use it?

File: includes/class-query-integration.php [hunk 2]
```diff
@@ -282,12 +280,26 @@ public function posts_clauses( $pieces, $query ) {
...
 	}
 
+	/**
+	 * Select all sticky listing ids and return thim in string comma seperated.
+	 *
+	 * @return string listing ids
+	 */
+	private function get_sticky_listing_ids() {
+		global $wpdb;
+		$results = $wpdb->get_results( "SELECT listing_id FROM {$wpdb->prefix}wpbdp_listings  WHERE is_sticky=1" );
```
Comment (stephywells) [inline includes/class-query-integration.php:295]: Let's run the query through our check_cache() functions too.

**Review by stephywells — CHANGES_REQUESTED** [review]
Thanks for trying that! I haven't tested it, but the code looks good, and that's great that it speeds it up a bit.

File: includes/class-query-integration.php (hunk 2 above)
Comment (stephywells) [reply includes/class-query-integration.php]: Thanks for the update!

The group here should be for the table getting searched. This helps us keep track of when the cache should be cleared.

`'group' => 'wpbdp_listings',`

**Review by stephywells — CHANGES_REQUESTED** [review]
Just left one more comment. This could use a final round of testing too since we made some updates.

File: includes/class-query-integration.php (hunk 2 above)
Comment (stephywells) [reply includes/class-query-integration.php]: I made this change and merged in my changes from the other pr.

**Review by stephywells — APPROVED** [review]

**Pull request merged**